Mathlib/NumberTheory/NumberField/ClassNumber.lean

```diff
@@ -45,6 +45,52 @@ theorem classNumber_eq_one_iff : classNumber K = 1 ↔ IsPrincipalIdealRing (rin
   card_classGroup_eq_one_iff
 #align number_field.class_number_eq_one_iff NumberField.classNumber_eq_one_iff
 
+open FiniteDimensional NumberField.InfinitePlace
+
+open scoped nonZeroDivisors Real
+
+theorem exists_ideal_in_class_of_norm_le (C : ClassGroup (𝓞 K)):
+    ∃ I : (Ideal (𝓞 K))⁰, ClassGroup.mk0 I = C ∧
+      Ideal.absNorm (I : Ideal (𝓞 K)) ≤ (4 / π) ^ NrComplexPlaces K *
+        ((finrank ℚ K).factorial / (finrank ℚ K) ^ (finrank ℚ K) * Real.sqrt |discr K|) := by
+  obtain ⟨J, hJ⟩ := ClassGroup.mk0_surjective C⁻¹
+  obtain ⟨_, ⟨a, ha, rfl⟩, h_nz, h_nm⟩ :=
+    exists_ne_zero_mem_ideal_of_norm_le_mul_sqrt_discr K (FractionalIdeal.mk0 K J)
+  obtain ⟨I₀, hI⟩ := Ideal.dvd_iff_le.mpr ((Ideal.span_singleton_le_iff_mem J).mpr (by convert ha))
+  have : I₀ ≠ 0 := by
+    contrapose! h_nz
+    rw [h_nz, mul_zero, show 0 = (⊥ : Ideal (𝓞 K)) by rfl, Ideal.span_singleton_eq_bot] at hI
+    rw [Algebra.linearMap_apply, hI, map_zero]
+  let I := (⟨I₀, mem_nonZeroDivisors_iff_ne_zero.mpr this⟩ : (Ideal (𝓞 K))⁰)
+  refine ⟨I, ?_, ?_⟩
+  · suffices ClassGroup.mk0 I = (ClassGroup.mk0 J)⁻¹ by rw [this, hJ, inv_inv]
+    exact ClassGroup.mk0_eq_mk0_inv_iff.mpr ⟨a, Subtype.ne_of_val_ne h_nz, by rw [mul_comm, hI]⟩
+  · rw [← FractionalIdeal.absNorm_span_singleton (𝓞 K), Algebra.linearMap_apply,
+      ← FractionalIdeal.coeIdeal_span_singleton, FractionalIdeal.coeIdeal_absNorm, hI, map_mul,
+      Nat.cast_mul, Rat.cast_mul, show Ideal.absNorm I₀ = Ideal.absNorm (I : Ideal (𝓞 K)) by rfl,
+      Rat.cast_coe_nat, Rat.cast_coe_nat, FractionalIdeal.coe_mk0,
+      FractionalIdeal.coeIdeal_absNorm, Rat.cast_coe_nat, mul_div_assoc, mul_assoc, mul_assoc]
+      at h_nm
+    refine le_of_mul_le_mul_of_pos_left h_nm ?_
+    exact Nat.cast_pos.mpr <| Nat.pos_of_ne_zero <| Ideal.absNorm_ne_zero_of_nonZeroDivisors J
+
+theorem classNumber_eq_one_of_abs_discr_lt
```

Comment:
Can you add the explicit fact that it is a PID?

Reply:
I changed the statement to 
```lean
theorem _root_.RingOfIntegers.isPrincipalIdealRing_of_abs_discr_lt
    (h : |discr K| < (2 * (π / 4) ^ NrComplexPlaces K *
      ((finrank ℚ K) ^ (finrank ℚ K) / (finrank ℚ K).factorial)) ^ 2) :
    IsPrincipalIdealRing (ringOfIntegers K) := by
```

Reply:
Great!